### PR TITLE
feat: drop support for PHP < 8.1

### PR DESCRIPTION
**Pull request by Chris53897**
https://github.com/Setono/CronExpressionBundle/issues/51

This is the minimal work to drop support for PHP < 8.1

### CODE REVIEW

**Comment by loevgaard** [comment]
Can you make this build or can you allow the static analysis to continue on error so that we can see the static analysis will be green for lower PHP versions than 8.4?

**Comment by Hanmac** [comment]
@loevgaard we need to update the vimeo/psalm dependency to minimum [5.24.0](https://github.com/vimeo/psalm/releases/tag/5.24.0) 
because of this commit: https://github.com/vimeo/psalm/commit/431977cb1d3f1562cd259c85e576bbd72c47f42e

**Comment by Chris53897** [comment]
Can the failing Static Code Analysis error be fixed in a follow up PR?
I do not know how to fix them at the moment.

**Comment by Hanmac** [comment]
> Can the failing Static Code Analysis error be fixed in a follow up PR? I do not know how to fix them at the moment.

Yeah, the Missing Template Param is already on my Radar

File: src/Form/TypeGuesser/CronExpressionTypeGuesser.php [hunk 1]
```diff
@@ -34,7 +35,6 @@ public function guessType($class, $property): ?TypeGuess
             return null;
         }
 
-
         $types = $this->extractor->getTypes($class, $property);
         if (!$types) {
```
Comment (Hanmac) [inline src/Form/TypeGuesser/CronExpressionTypeGuesser.php:39]: @Chris53897 before I make an extra PR, use `if (is_null($types)) {` to fix RiskyTruthyFalsyComparison

**Comment by Hanmac** [comment]
@Chris53897 

As for the MissingTemplateParam:


 CronExpressionToPartsTransformer needs:
```php
/**
 * @template-implements DataTransformerInterface<CronExpression, array<int>>
 */

```

 CronExpressionToStringPartsTransformer needs:
```php
/**
 * @template-implements DataTransformerInterface<CronExpression, array<string>>
 */

```

 CronExpressionToStringTransformer needs:
```php
/**
 * @template-implements DataTransformerInterface<CronExpression, string>
 */

```

CronExpressionTest needs:
```php
/**
 * @template-implements ConstraintValidatorTestCase<CronExpressionValidator>
 */
```

FormCallbackTest needs:
```php
/**
 * @template-implements ConstraintValidatorTestCase<CallbackValidator>
 */
```

**Comment by Chris53897** [comment]
@Hanmac Thanks. I made your changes. But i still got errors

I would suggest to merge this at it is. And make Psalm happy in a follow up PR

**Comment by Hanmac** [comment]
arg the template params are tricky :(

symfony 6+ wants template params
symfony 5+ doesn't want template params

also i manually run the symfony6 tests:
> Error: src/Form/DataTransformer/CronExpressionToPartsTransformer.php:20:40: LessSpecificImplementedReturnType: The inherited return type 'array<array-key, int>|null' for Symfony\Component\Form\DataTransformerInterface::transform is more specific than the implemented return type for Setono\CronExpressionBundle\Form\DataTransformer\CronExpressionToPartsTransformer::transform 'array<array-key, mixed>' (see https://psalm.dev/166)
Error: src/Form/DataTransformer/CronExpressionToStringPartsTransformer.php:20:40: LessSpecificImplementedReturnType: The inherited return type 'array<array-key, string>|null' for Symfony\Component\Form\DataTransformerInterface::transform is more specific than the implemented return type for Setono\CronExpressionBundle\Form\DataTransformer\CronExpressionToStringPartsTransformer::transform 'array<array-key, mixed>' (see https://psalm.dev/166)

**Comment by Hanmac** [comment]
The Tests need `@template-extends` instead of `template-implements`

File: src/Form/TypeGuesser/CronExpressionTypeGuesser.php (hunk 1 above)
Comment (Hanmac) [reply src/Form/TypeGuesser/CronExpressionTypeGuesser.php]: What I hate there, is that the code-style fixers can't keep straight if it should be `is_null` or `null ===` :(

**Comment by Chris53897** [comment]
I think the solution is to just run psalm on a specific test-run. Not for every combination.
Like PHP 8.3 and Smyfony 7.1 as highest.

**Comment by Hanmac** [comment]
@loevgaard @Chris53897 finally! All Green again!

**Comment by loevgaard** [comment]
Good job! You can just merge when you find it acceptable, @Hanmac :-)

**Comment by Hanmac** [comment]
@loevgaard it says "Review required", so you need to give your 👍, and might merge-squash yourself

**Comment by loevgaard** [comment]
I think you can review and merge, can't you? :)

**Review by Hanmac — APPROVED** [review]

**Comment by Hanmac** [comment]
it works when I am not the one that created the issue

**Comment by loevgaard** [comment]
Yeah! Great! Thanks 🎉

**Pull request merged**